lib/ansible/modules/cloud/azure/azure_rm_securitygroup.py

```diff
@@ -343,86 +349,68 @@ def validate_rule(self, rule, rule_type=None):
     :param rule_type: Set to 'default' if the rule is part of the default set of rules.
     :return: None
     '''
-
-    if not rule.get('name'):
-        raise Exception("Rule name value is required.")
-
-    priority = rule.get('priority', None)
-    if not priority:
-        raise Exception("Rule priority is required.")
-    if not isinstance(priority, integer_types):
-        try:
-            priority = int(priority)
-            rule['priority'] = priority
-        except:
-            raise Exception("Rule priority attribute must be an integer.")
+    priority = rule.get('priority', 0)
     if rule_type != 'default' and (priority < 100 or priority > 4096):
         raise Exception("Rule priority must be between 100 and 4096")
 
-    if not rule.get('access'):
-        rule['access'] = 'Allow'
-
-    access_names = [member.value for member in self.nsg_models.SecurityRuleAccess]
-    if rule['access'] not in access_names:
-        raise Exception("Rule access must be one of [{0}]".format(', '.join(access_names)))
-
-    if not rule.get('destination_address_prefix'):
-        rule['destination_address_prefix'] = '*'
-
-    if not rule.get('source_address_prefix'):
-        rule['source_address_prefix'] = '*'
+    def check_plural(src, dest):
+        if isinstance(rule.get(src), list):
+            rule[dest] = rule[src]
+            rule[src] = None
 
-    if not rule.get('protocol'):
-        rule['protocol'] = '*'
+    check_plural('destination_address_prefix', 'destination_address_prefixes')
+    check_plural('source_address_prefix', 'source_address_prefixes')
+    check_plural('source_port_range', 'source_port_ranges')
+    check_plural('destination_port_range', 'destination_port_ranges')
 
-    protocol_names = [member.value for member in self.nsg_models.SecurityRuleProtocol]
-    if rule['protocol'] not in protocol_names:
-        raise Exception("Rule protocol must be one of [{0}]".format(', '.join(protocol_names)))
 
-    if not rule.get('direction'):
-        rule['direction'] = 'Inbound'
-
-    direction_names = [member.value for member in self.nsg_models.SecurityRuleDirection]
-    if rule['direction'] not in direction_names:
-        raise Exception("Rule direction must be one of [{0}]".format(', '.join(direction_names)))
-
-    if not rule.get('source_port_range'):
-        rule['source_port_range'] = '*'
+def compare_rules_change(old_list, new_list, purge_list):
+    old_list = old_list or []
+    new_list = new_list or []
+    changed = False
 
-    if not rule.get('destination_port_range'):
-        rule['destination_port_range'] = '*'
+    for old_rule in old_list:
+        matched = next((x for x in new_list if x['name'] == old_rule['name']), [])
+        if matched:  # if the new one is in the old list, check whether it is updated
```

Comment:
the code is ok here, but I did different solution for this comparison recently, first convert to dictionary and then compare:
        oldd = {}
        for item in old:
            name = item['name']
            oldd[name] = item
        newd = {}
        for item in new:
            name = item['name']
            newd[name] = item
        return  newd == oldd

I just wonder what you think? No need to update

Reply:
oh, not formatted correctly, anyway i think it's pretty simple

Reply:
There are something different, here also construct the new parameter, since there is a `purge_list` flag. My experience to compare two list of dict is 

```python
def compare_list_of_dict(A, B):
    def hash_dict(item):
        return str(dict(
            name=to_native(item.name),  # str
            id=item.id,  # int
            is_flag=item.is_flag  # boolean
            ...
        ))
   return set(hash_dict(A)) == set(hash_dict(B))
```